### ISSUE TITLE
[MENFORCER-447] Verify working with Maven 4

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-enforcer#818]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** opened **[MENFORCER-447](https://issues.apache.org/jira/browse/MENFORCER-447?redirect=false)** and commented

---

**Issue Links:**
- [MNG-7228](https://issues.apache.org/jira/browse/MNG-7228) MavenProject.getOriginalModel returns interpolated parts
 (_**"is blocked by"**_)
- [MNG-7648](https://issues.apache.org/jira/browse/MNG-7648) Generated model reader is not setting location information
 (_**"is blocked by"**_)

**Remote Links:**
- [GitHub Pull Request #224
](https://github.com/apache/maven-enforcer/pull/224)
- [GitHub Pull Request #225
](https://github.com/apache/maven-enforcer/pull/225)

**Backported to:** [3.2.0](https://github.com/apache/maven-enforcer/milestone/21?closed=1)

**Comment by jira-importer** [comment]
**[Pavel Horal](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=pavel.horal)** commented

The rule `requirePluginVersions` is not able to pick up versions specified in the parent POM. The issue is demonstrated in my test repository https://github.com/pavelhoral/enforcer-plugin-test .

I have tried to find the issue and so far got to this call that incorrectly returns empty optional - [PluginWrapper.java#L75](https://github.com/apache/maven-enforcer/blob/374d7d6a98fb9a33e5740936d612eadfce9cbb66/enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/utils/PluginWrapper.java#L75) (called from here [RequirePluginVersions.java#L881](https://github.com/apache/maven-enforcer/blob/374d7d6a98fb9a33e5740936d612eadfce9cbb66/enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/RequirePluginVersions.java#L881)). By the way the JavaDoc of that method is not correct as empty optional is not the same as null result.

**Update 1:** Plugin model entity does not have `locations` property set. Without that enforcer is not able to tell where the plugin is defined.

**Update 2:** The issue seems to be that `MavenXpp3Reader#parsePlugin` is not setting location. So the issue is probably in Maven Core / Maven Model, not in Enforcer - https://github.com/apache/maven/blob/e73a0b00fde80c400a6d854ec0c2ba7436325eed/maven-toolchain-model/src/main/mdo/reader.vm#L683

**Update 3:** Created MNG-7648

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

Rules not needed for Maven 4:
- banDuplicatePomDependencyVersions
- requireProfileIdsExist

Maven 4 strict checking such rules, so build will not start with projects contains such violations.
We need skip testing those rules with Maven 4 - in order to build pass.

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

`requireNoRepositories` is affected by MNG-7228

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

Other rules are ok with Maven 4

**Comment by jira-importer** [comment]
**[Slawomir Jaranowski](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=slawekjaranowski)** commented

I added workaround in `requireNoRepositories` for repositories from `settings.xml` - now all IT test pass with the latest version of Maven 4
